test/unit/message/outgoing-message-test.js

```diff
@@ -0,0 +1,186 @@
+
+    const packets = [];
+
+    message.on('data', function(packetData) {
+      const packet = new Packet(packetData);
+
+      packets.push(packet);
+    });
+
+    message.write(data);
+    message.end(function() {
+      test.strictEqual(3, packets.length);
+
+      test.strictEqual(1, packets[0].packetId());
+      test.strictEqual(1, packets[0].type());
+      test.ok(!packets[0].isLast());
+      test.strictEqual(20, packets[0].length());
+      test.deepEqual(Buffer.from('123456789|12'), packets[0].data());
+
+      test.strictEqual(2, packets[1].packetId());
+      test.strictEqual(1, packets[1].type());
+      test.ok(!packets[1].isLast());
+      test.strictEqual(20, packets[1].length());
+      test.deepEqual(Buffer.from('3456789|1234'), packets[1].data());
+
+      test.strictEqual(3, packets[2].packetId());
+      test.strictEqual(1, packets[2].type());
+      test.ok(packets[2].isLast());
+      test.strictEqual(13, packets[2].length());
+      test.deepEqual(Buffer.from('56789'), packets[2].data());
+
+      test.done();
+    });
+  },
+
+  'transforms the data as soon as possible': function(test) {
+    const debug = new Debug();
+    const message = new OutgoingMessage(1, false, 20, debug);
+    const data = Buffer.from('123456789|123456789|123456789');
+
+    const packets = [];
+
+    message.on('data', function(packetData) {
+      const packet = new Packet(packetData);
+
+      packets.push(packet);
+    });
+
+    message.write(data, function() {
+      test.strictEqual(2, packets.length);
+
+      test.strictEqual(1, packets[0].packetId());
+      test.strictEqual(1, packets[0].type());
+      test.ok(!packets[0].isLast());
+      test.strictEqual(20, packets[0].length());
+      test.deepEqual(Buffer.from('123456789|12'), packets[0].data());
+
+      test.strictEqual(2, packets[1].packetId());
+      test.strictEqual(1, packets[1].type());
+      test.ok(!packets[1].isLast());
+      test.strictEqual(20, packets[1].length());
+      test.deepEqual(Buffer.from('3456789|1234'), packets[1].data());
+    });
+
+    message.end(function() {
+      test.strictEqual(3, packets.length);
+
+      test.strictEqual(1, packets[0].packetId());
+      test.strictEqual(1, packets[0].type());
+      test.ok(!packets[0].isLast());
+      test.strictEqual(20, packets[0].length());
+      test.deepEqual(Buffer.from('123456789|12'), packets[0].data());
+
+      test.strictEqual(2, packets[1].packetId());
+      test.strictEqual(1, packets[1].type());
+      test.ok(!packets[1].isLast());
+      test.strictEqual(20, packets[1].length());
+      test.deepEqual(Buffer.from('3456789|1234'), packets[1].data());
+
+      test.strictEqual(3, packets[2].packetId());
+      test.strictEqual(1, packets[2].type());
+      test.ok(packets[2].isLast());
+      test.strictEqual(13, packets[2].length());
+      test.deepEqual(Buffer.from('56789'), packets[2].data());
+
+      test.done();
+    });
+  },
```

Comment:
You may add one corner case test - data size multiple of packet size so that the last packet is a full packet.